### ISSUE TITLE
ETIMEOUT (Ping timeout) does not autoreconnect

### ISSUE THREAD

**Issue by davibe** [kiwiirc/irc-framework#64]
I am debugging thelounge irc client. Looking at the irc-framework code i don't understand how a socket error like ETIMEOUT would trigger reconnect. What i see happening is that this gets called

https://github.com/kiwiirc/irc-framework/blob/842fa112d5f0eda87409b7c95d8b36b14a4573f8/src/connection.js#L281

which sets `requested_disconnect` to true, which in turn makes autoreconnect not work.

How was it intended to work ?

**Comment by davibe** [comment]
Here i am testing something

https://github.com/davibe/irc-framework

**Comment by davibe** [comment]
No feedback after so many months ?

**Comment by davibe** [comment]
this is the same as #56

**Comment by prawnsalad** [comment]
Sorry for the late reply, has been a busy few months :)

Your original issue with ETIMEOUT is not the same as you've changed it to. Ping Timeout occurs when the client has successfully connected to the IRC network and then does not reply to a PING at some point.

I've purposely got it to only reconnect if it had already successfully connected at least once before otherwise it's a waste of time. If you get disconnected, and then it gets connection errors on reconnect, it will attempt to reconnect until it hits the max reconnection numbers.

**Comment by davibe** [comment]
Ping timeout happens here
https://github.com/kiwiirc/irc-framework/blob/842fa112d5f0eda87409b7c95d8b36b14a4573f8/src/client.js#L231

When there is a ping timeout you do `quit()` which is here
https://github.com/kiwiirc/irc-framework/blob/842fa112d5f0eda87409b7c95d8b36b14a4573f8/src/client.js#L282

then it calls `end()` which is here
https://github.com/kiwiirc/irc-framework/blob/842fa112d5f0eda87409b7c95d8b36b14a4573f8/src/connection.js#L274

then it calls `disposeSocket()` here
https://github.com/kiwiirc/irc-framework/blob/842fa112d5f0eda87409b7c95d8b36b14a4573f8/src/connection.js#L228

as i read it : "if there was a socket and we were connected then this disconnection was requested explicitly therefore we are not going to try to reconnect"

I think this is wrong. In case of a ping timeout, if we were connected before, we should try to reconnect again. 

Am i reading it wrong ?

The first time I wrote the original post i was debugging the very same thing. I called it ETIMEOUT because i read it in the log. I guess ping timeout and ETIMEOUT happen together quite often as they both are caused by network/server/tcp problems. I have been experiencing this issue while using `thelounge` irc client. My workaround was to comment out `that.quit(..)` when there is a ping timeout. Somehow that makes it go on with the connection attempts, but it's not clean.
